### PR TITLE
feat: Add the FDv2 data system orchestrator

### PR DESCRIPTION
**Pull request by beekld**
## Summary

Adds the FDv2 orchestrator, the `DataSystem` implementation that owns the in-memory store and keeps it populated. It runs an initializer phase to obtain a basis, then a synchronizer phase for ongoing changes.

The initializer phase tries each initializer in order until one produces a basis. The synchronizer phase then runs one synchronizer at a time and rotates between them on three triggers:

- A fallback timer advances to the next synchronizer after a sustained interruption.
- A recovery timer returns to the primary synchronizer after the recovery interval.
- A terminal error blocks a synchronizer and advances past it.

Initialization is signaled as complete on the first basis and as failed once every source is exhausted without one, because the client's initialization path already resolves to success or failure. Synchronizer status transitions are logged, with interruptions at info (not repeated while ongoing) and terminal errors at warn.

---

> [!NOTE]
> **Overview**
> Introduces **`FDv2DataSystem`**, the FDv2 **`DataSystem`** implementation that owns an in-memory store and keeps it updated via a two-phase async **`run`** loop spawned from **`start`**.
> 
> **Initializer phase** walks factory-built initializers until one returns a **basis** (full changeset with a selector), applying each non-`None` changeset and signaling **`init_complete(true)`** once any full payload lands (after the initializer chain when needed). **Synchronizer phase** drives one synchronizer at a time through **`SourceManager`**: cyclic rotation, blocking on terminal errors, **fallback** after sustained **`Interrupted`**, and **recovery** back to the prime when a backup source was active. Selectors are passed into **`next`** and updated per changeset rules (`None` kind preserves selector; selectorless partial clears it).
> 
> Wires the module in **`fdv2/mod.rs`** and drops **`dead_code`** allowances on **`ChangeSet::selector`** and **`TransactionalDataStore::apply`** now that the orchestrator consumes them. Adds a large **`#[cfg(test)]`** suite covering init/failure, rotation, timers, and selector edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f8ccaf35e4b817c3ccf1a8609e7cd4cdf29496a8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 1]
```diff
@@ -0,0 +1,913 @@
...
+                        selector = change_set.selector.clone();
+                        store.write().apply(change_set);
+                        if !initialized {
+                            init_complete(true);
+                            initialized = true;
+                        }
+                        // Good data clears the fallback countdown.
+                        fallback_at = None;
+                        interrupted_logged = false;
+                    }
```
[the remark this answers is not in the document]
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs:251]: Fixed.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+    }
+}
+
+impl DataSystem for FDv2DataSystem {
+    fn start(
+        &self,
+        init_complete: Arc<dyn Fn(bool) + Send + Sync>,
+        shutdown_receiver: broadcast::Receiver<()>,
+    ) {
+        let initializers = self
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs]: If the first initializer returns a basis, then we wouldn't need to build the subsequent ones. Maybe we should build them as we iterate instead of all at once then?

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 3]
```diff
@@ -0,0 +1,952 @@
...
+        self.synchronizer_index = None;
+    }
+
+    /// Whether the active factory is the first available one.
+    fn is_prime(&self) -> bool {
+        let first = self
+            .states
+            .iter()
+            .position(|s| *s == SourceState::Available);
+        matches!((first, self.current_factory_index), (Some(f), Some(c)) if f == c)
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs]: Can this not be simplified down first == self.current_factory_index && first.is_some()?

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+
+    // Initializer phase: try each in order until one yields a basis.
+    for mut initializer in initializers {
+        let name = initializer.name().to_string();
+        let mut shutdown = Box::pin(shutdown_receiver.recv()).fuse();
+        futures::select! {
+            _ = shutdown => return,
+            event = initializer.run().fuse() => {
+                if let FDv2SourceResult::ChangeSet(change_set) = event.result {
+                    if !matches!(change_set.kind, ChangeSetKind::None) {
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs]: Initializers can arrive with a full change set but no basis (e.g. file based initialization). So applying a change set and determining if the init chain is done are two different things.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+    }
+
+    fn event(result: FDv2SourceResult) -> FDv2SourceEvent {
+        FDv2SourceEvent {
+            result,
+            fdv1_fallback: None,
+        }
+    }
+
+    struct MockInitializer {
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs:338]: Might be worth implementing the test data source for FDv2 so future tests can rely on that if needed.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 6]
```diff
@@ -0,0 +1,952 @@
...
+        notify.notified().await;
+
+        // The basis was applied and is readable through the store() handle.
+        assert_eq!(*calls.lock().unwrap(), vec![true]);
+        assert!(system.store().read().flag("f1").is_some());
+        drop(shutdown_tx);
+    }
+
+    #[tokio::test]
+    async fn initializer_basis_signals_once_and_propagates_selector() {
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs:506]: Based on my earlier comment, you will also want a test that gets a payload from the first source, doesn't have a basis, and then continues onto the second one.

**Review by keelerm84 — COMMENTED** [review]
A few questions and suggestions on the orchestrator, mostly around the initializer chain and test coverage.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 2 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: Fair point. Done.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 3 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: Done.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 4 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: Ah, yeah, I guess I was assuming any initializer that finished would set a basis, but that's not a good assumption. Fixed. And test added.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 6 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs:506]: Test added.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 5 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs:338]: I'm not convinced that would actually make these tests any simpler, because we're mocking out the initializers and synchronizers to make them behave in non-standard ways to exercise the orchestrator paths.

But I do have the test data source in my road map, and I intend to do it in a later PR.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 7]
```diff
@@ -178,22 +174,32 @@ async fn run(
...
+                            selector = change_set.selector.clone();
+                        }
                         store.write().apply(change_set);
-                        init_complete(true);
-                        initialized = true;
-                        break;
+                        if !initialized {
+                            init_complete(true);
+                            initialized = true;
+                        }
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs]: [We should only mark this as initialized if we received the basis.](https://github.com/launchdarkly/sdk-specs/blob/be4d547a9071c083d5211d5e874f8188d074dab8/specs/DATASYSTEM-data-system/v2/README.md?plain=1#L709)

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 8]
```diff
@@ -226,7 +232,9 @@ async fn run(
                 event = next => match event.result {
                     FDv2SourceResult::ChangeSet(change_set) => {
                         if !matches!(change_set.kind, ChangeSetKind::None) {
-                            selector = change_set.selector.clone();
+                            if change_set.selector.is_some() {
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs]: [Same deal here.](https://github.com/launchdarkly/sdk-specs/blob/be4d547a9071c083d5211d5e874f8188d074dab8/specs/DATASYSTEM-data-system/v2/README.md?plain=1#L710)

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 9]
```diff
@@ -570,20 +626,50 @@ mod tests {
...
+            store,
+            init_complete,
+            shutdown_rx,
+            FALLBACK_TIMEOUT,
+            RECOVERY_TIMEOUT,
+        )
+        .await;
+
+        // The request after the selectorless change still carries "s1".
+        let seen = selectors_seen.lock().unwrap();
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs:745]: I don't think this is right.

If we maintain a basis even after a subsequent change, then we can't safely pick back up where we left off because we have a different state than the basis suggests to FD.

**Review by keelerm84 — CHANGES_REQUESTED** [review]

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 4 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: Okay, so I've made it so that `init_complete` is called as soon as there's a full payload, which is consistent with Java and C++ (and FDv1). I don't think calling `init_complete` should require a selector, nor does it necessarily have to happen from an initializer. We may not get a full payload until we get to a synchronizer, if the initializers fail. And if we're on FDv1 fallback, we may get usable data without a selector.

And I've fixed it so that we don't exit the initializer loop until we get a selector. I think we're all on the same page about that.

If that doesn't sound right to you, then we should probably have a quick chat about it.

Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: (or at least, that will be the case after i finish pushing)

Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: Okay, pushed.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 7 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: When you say "basis" here, do you mean (a) a full payload, or (b) a full payload with a selector?

As mentioned above, based on my interpretation of the spec, and the C++ and Java implementations, I think we want to mark as initialized as soon as we get a full payload, regardless of selector, because that's when the data store becomes usable. But we don't advance to the synchronizers until we get the selector.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 8 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs]: Agreed that this should be consistent with the other spot.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 9 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs:745]: You're right, and this is an existing bug in C++ as well. Fixed, and added a test.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 10]
```diff
@@ -0,0 +1,1223 @@
...
+
+use crate::data_system::DataSystem;
+use crate::stores::store::{DataStore, InMemoryDataStore, TransactionalDataStore};
+
+use super::model::{ChangeSetKind, Selector};
+use super::source::{FDv2SourceResult, Initializer, Synchronizer};
+
+/// Produces a fresh initializer each time the orchestrator starts a run.
+pub(crate) trait InitializerFactory: Send + Sync {
+    fn create(&self) -> Box<dyn Initializer>;
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs:17]: Didn't notice this the first round through, but I think the convention within this repository tends to be more "Builder / build" than "Factory / create". Do you think we should align so it's more consistent for the customer? (Not going to block over this of course)

File: launchdarkly-server-sdk/src/fdv2/data_system.rs [hunk 11]
```diff
@@ -0,0 +1,1223 @@
...
+    for factory in initializer_factories {
+        let mut initializer = factory.create();
+        let name = initializer.name().to_string();
+        let mut shutdown = Box::pin(shutdown_receiver.recv()).fuse();
+        futures::select! {
+            _ = shutdown => return,
+            event = initializer.run().fuse() => {
+                match event.result {
+                    FDv2SourceResult::ChangeSet(change_set) => {
+                        let is_full = matches!(change_set.kind, ChangeSetKind::Full);
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/data_system.rs:188]: I think there is still a difference between this and the Java implementation. The java implementation does not require a full change set. If ANY change set is received, it considers that initialized. So a partial change set with a basis would behave differently between the two. I _think_ the Java one will also trigger if it receives an "up-to-date" without any payload data as well.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 10 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs:17]: > Do you think we should align so it's more consistent for the customer? 

This is not a public API. It's an internal factory for creating initializer instances. The public config API comes in a later PR and does use the "Builder / build" terminology to build the config.

File: launchdarkly-server-sdk/src/fdv2/data_system.rs (hunk 11 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/data_system.rs:188]: I'm happy to make this work like Java, but I'm a little confused.
1. Is it even possible to get a partial result from a request that didn't already have a basis?
2. Is it possible to get a partial result before any full result?

I was assuming that partial results are always diffs from an existing basis, but these cases where Rust and Java diverge are only cases where there's a partial result before any basis.

**Review by keelerm84 — APPROVED** [review]
